### ISSUE TITLE
"wait: true" doesn't wait and throws event "ReconciliationSucceeded" before resources are deployed

### ISSUE THREAD

**Issue by alexander-matthiesen** [fluxcd/kustomize-controller#653]
We are currently migrating out services to GitOps with FluxCD. In this process we are quite struggeling with a few things but one of the main blockers for us is currently the "not waiting" of the kustomize-controller. 

We have a multi-tenant setup and all of our microservices have an application repository and an environment repository. The application repository contains source code, tests and documentation. The environment repository contains the environment configuration for different stages (with differentiation by folders and not by branches).

Regarding our workflow and to give some context why we need the "wait":

1. The developer pushes into the application repository branch
2. DockerImage, HelmChart and other artifacts are build in the app repo
3. A GitLab job is writing the current version in the env repo in the folder "stages/dev"
4. (Flux will monitor that change and deploy the application)
5. A custom controller (pipeline-trigger) will watch the tenant-kustomization and is looking for "ReconciliationSucceeded"
6. The pipeline-trigger will trigger the application repository and execute e2e tests against the deployed application
7. The flow will continue with step 3 but for stages "INT, QA and PROD"

Currently the issue is, that step 6 is directly executed after the GitRepository has triggered the reconciliation.

So the questions for me are:

- Do we monitor the wrong events?
- Is FluxCD not waiting at all?
- What are the best configs for the interval settings together with "wait"?

Thanks in advance!

**Comment by haarchri** [comment]
@alexander-matthiesen can you check for healthy ? We had a short discussion at KubeCon with @stefanprodan about this

**Comment by alexander-matthiesen** [comment]
@haarchri you mean the event "healthy"? I will test a few things and come back here.

**Comment by stefanprodan** [comment]
Yes there is a dedicated condition in the status named `Healthy` and an event is issued to reflect the state.

**Comment by alexander-matthiesen** [comment]
Thanks for the fast reply! 

So currently only the condition is set to Healthy but we are getting no event in the kustomization-controller. Only "ReconciliationSucceeded", "HealthCheckFailed" or "ReconciliationFailed". 

Does the "Healthy" condition is also triggering an event?

In rare occurrences the kind "HelmChart" having the status "ArtifactFailed". Will this also be reflected in the kustomization-controller "HealchCheckFailed" or do we have to observe those events / alerts separately? 

Current we are using the [notification-controller](https://fluxcd.io/docs/components/notification/alert/#example) functionality to determine when to trigger out post-deploy-pipeline (which should also notify the developers when something bad happens).

**Comment by stefanprodan** [comment]
You can see the logic here https://github.com/fluxcd/kustomize-controller/blob/main/controllers/kustomization_controller.go#L886

To avoid spam, we only send a healthy event if the previous reconciliation had HealthCheckFailed

**Comment by haarchri** [comment]
Which Version of Flux is installed in this Clusters ?

**Comment by alexander-matthiesen** [comment]
Ah, the last part is important "If the source revision is newer". Thats good to hear.

We are using the version 0.24.0 of fluxcd in our "intermediate" cluster.

**Comment by alexander-matthiesen** [comment]
@stefanprodan , is it really an event with the reason "Healthy" or is it the condition that is set? Currently we are monitoring only the reason of the events and not the condition of certain controllers / resources.

**Comment by alexander-matthiesen** [comment]
The reason for the health check passed event is not "Healthy" but "Progressing".

(I've changed the received message and used logfmt)

`2022-06-01T18:50:55.181570300Z INFO:app:controller=kustomize-controller namespace=flux-system kind=Kustomization name=service-alpha-api-dev reason=Progressing message=Health check passed in 30.0546639s`

**Comment by alexander-matthiesen** [comment]
Okay, yes I got it, that there are reasons and conditions. 

Unfortunately it seems that the ReconciliationSucceeded event is also thrown when there is no new revision but the kustomization-controller has successfully reconciled (So every interval that you set in the kustomization resources). Is that true?

So since there is a HealthCheckFailed reason, is there any disadvantage to introduce a HealthCheckPassed reason which is thrown when the condition changed to "Healthy" again? In other words: Which is thrown when the revision changed and the HealthCheck actually passed?

This would enable us to set "wait: true" and just monitoring the "Reason"-Field and check if it was HealthCheckFailed or HealthCheckPassed.

**Comment by stefanprodan** [comment]
@alexander-matthiesen you could filter for messages beginning with `Health check passes`

**Comment by sch1ldkr0ete** [comment]
I just came across this thread as we are using Flux to dynamically create preview namespaces from kustomizations. We set the version for the image using variable substitution. Now we would like to be notified when a new environment is successfully created. According to this thread and [L859](https://github.com/fluxcd/kustomize-controller/blob/814a79f8c127dd23be1ed476729d04ee8516a41a/controllers/kustomization_controller.go#L859), that doesn't seem to be possible since we don't create a new revision in the source with variable substitution.

Is there any way to be notified in these cases?
